source/common/config/delta_subscription_impl.h

```diff
@@ -0,0 +1,223 @@
+#pragma once
+
+#include <queue>
+
+#include "envoy/api/v2/discovery.pb.h"
+#include "envoy/common/token_bucket.h"
+#include "envoy/config/subscription.h"
+
+#include "common/common/assert.h"
+#include "common/common/backoff_strategy.h"
+#include "common/common/logger.h"
+#include "common/common/token_bucket_impl.h"
+#include "common/config/grpc_stream.h"
+#include "common/config/utility.h"
+#include "common/grpc/common.h"
+#include "common/protobuf/protobuf.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Config {
+
+struct ResourceNameDiff {
+  std::vector<std::string> added_;
+  std::vector<std::string> removed_;
+};
+
+/**
+ * Manages the logic of a (non-aggregated) delta xDS subscription.
+ */
+template <class ResourceType>
+class DeltaSubscriptionImpl
+    : public Subscription<ResourceType>,
+      public GrpcStream<envoy::api::v2::DeltaDiscoveryRequest,
+                        envoy::api::v2::DeltaDiscoveryResponse, ResourceNameDiff> {
+public:
+  DeltaSubscriptionImpl(const LocalInfo::LocalInfo& local_info, Grpc::AsyncClientPtr async_client,
+                        Event::Dispatcher& dispatcher,
+                        const Protobuf::MethodDescriptor& service_method,
+                        Runtime::RandomGenerator& random, Stats::Scope& scope,
+                        const RateLimitSettings& rate_limit_settings, SubscriptionStats stats)
+      : GrpcStream<envoy::api::v2::DeltaDiscoveryRequest, envoy::api::v2::DeltaDiscoveryResponse,
+                   ResourceNameDiff>(std::move(async_client), service_method, random, dispatcher,
+                                     scope, rate_limit_settings),
+        type_url_(Grpc::Common::typeUrl(ResourceType().GetDescriptor()->full_name())),
+        local_info_(local_info), stats_(stats) {
+    request_.set_type_url(type_url_);
+    request_.mutable_node()->MergeFrom(local_info_.node());
+  }
+
+  // Enqueues and attempts to send a discovery request, (un)subscribing to resources missing from /
+  // added to the passed 'resources' argument, relative to resources_. Updates resources_ to
+  // 'resources'.
+  void buildAndQueueDiscoveryRequest(const std::vector<std::string>& resources) {
+    ResourceNameDiff diff;
+    std::set_difference(resources.begin(), resources.end(), resource_names_.begin(),
+                        resource_names_.end(), std::inserter(diff.added_, diff.added_.begin()));
+    std::set_difference(resource_names_.begin(), resource_names_.end(), resources.begin(),
+                        resources.end(), std::inserter(diff.removed_, diff.removed_.begin()));
+
+    for (const auto& added : diff.added_) {
+      resources_[added] = "0";
```

Comment:
?

Reply:
Oh, this was when I didn't realize that version is supposed to be utterly opaque to xDS. Thanks for pointing it out. I think empty string should work here.